Fix crash on kmeans2

vbarrielle opened the pull request:
scipy.cluster.vq.kmeans2 would crash on random initialization when its data argument contains less data points than dimensions.

This pull request adds a test, test_kmeans2_high_dim, which highlights the issue, and fixes the random initialization procedure to use pca instead of cholesky factorization of the covariance matrix when the number of data points is less than the dimension.

ev-br commented:
Looks reasonable. I confirm both the LinAlgError and that this PR fixes it.
I'll keep it open for a while in case someone has comments. Otherwise, it's a bugfix I'm going to merge soon.

pv commented on scipy/cluster/vq.py:
```diff
@@ -634,9 +634,20 @@ def init_rankn(data):
+        mu = np.mean(data, axis=0)
+        _, s, vh = np.linalg.svd(data - mu, full_matrices=False)
+        x = np.random.randn(k, s.size)
+        sVh = s[:,None] * vh
```
I think you need to divide by `sqrt(n-1)`

vbarrielle replied:
Indeed, I had forgotten about it.

pv commented on scipy/cluster/tests/test_vq.py:
```diff
@@ -214,6 +214,13 @@ def test_kmeans2_rank1_2(self):
+        # of input points
+        data = np.fromfile(DATAFILE1, sep=", ")
+        data = data.reshape((20, 20))[:10]
+        kmeans2(data, 2)
```
There should also be a test checking the produced result has the correct covariance.
IOW, use a large `k`, set `np.random.seed(1234)`, and compare the result from `np.cov(x, rowvar=0)` to `np.cov(data, rowvar=0)`

vbarrielle commented:
I've fixed the missing division, and added a test to check the covariance of the generated data. The comparison of the covariances makes use of a relatively loose tolerance but I can't get better precision without using a much larger `k`, which uses too much memory IMO.

ev-br commented:
`np.fromfile` usage needs to be removed from tests, now that https://github.com/scipy/scipy/pull/5573 is merged. Here is the branch where I did this,  squashed the commits and rebased on master 
https://github.com/ev-br/scipy/commits/pr/5486

An unpleasant thing is that this line, `x = np.dot(x, sVh) + mu`, fails with a MemoryError in a wine VM (which, admittedly, is rather tight on memory).

vbarrielle commented:
> An unpleasant thing is that this line, `x = np.dot(x, sVh) + mu`, fails with a MemoryError in a wine VM (which, admittedly, is rather tight on memory). 
The MemoryError surely comes from the covariance test. It needs quite a lot of data to get a nice approximation of the covariance matrix. Maybe the test ought to be disabled on platforms without enough memory?

> `np.fromfile` usage needs to be removed from tests, now that #5573 is merged. Here is the branch where I did this, squashed the commits and rebased on master

Is there something I should do about this? I'm unfamiliar with the process, should I update my pull request using your branch, or are you going to merge your branch?

ev-br commented:
If you have an idea of how to trigger the problem for smaller problem size, it'd be helpful (eg, do you really need k=1e6?). If you can do it, please push a commit to your branch, I'll cherry-pick it. Otherwise, I'll just add a skip condition and merge the rebased one.

vbarrielle commented:
> If you have an idea of how to trigger the problem for smaller problem size,
> it'd be helpful (eg, do you really need k=1e6?). If you can do it, please
> push a commit to your branch, I'll cherry-pick it. Otherwise, I'll just add
> a skip condition and merge the rebased one.

The `k=1e6` is only required to test that the _krandinit actually produces the 
correct covariance in the high dimensional case. With lower values it's harder 
to see if the obtained covariance matches the initial covariance.

However, this test is only here to check the correctness of the fix, and as 
such it is only required to pass on platforms with enough memory.

The `test_kmeans2_high_dim`, on the other hand, actually highlights the bug, 
and should pass on any platform.

So I guess adding a skip condition is the way to go.

ev-br commented:
Rebased, squashed and merged with the skip condition on the test in https://github.com/scipy/scipy/commit/db14b78e711bdb7c9611160f9d476cc936310433. Thanks @vbarrielle.
